doc/ml_features.rst

```diff
@@ -0,0 +1,337 @@
+
+For details about ``set_model_from_file``, please
+refer to :ref:`SmartRedis API <smartredis-api>`.
+Finally, the model can be run on the sample and the output is ready to be downloaded.
+
+.. code-block:: python
+
+    client.run_model(model_key, "mnist_input", "mnist_output")
+    pred = client.get_tensor("mnist_output")
+
+
+If storing the model as a file is not needed, then it can just be kept in memory
+after serialization, using ``serialize_model`` after compiling the model. The same
+workflow we saw in the previous example can then basically be achieved by replacing
+``set_model_from_file`` with ``set_model``:
+
+.. code-block:: python
+
+    # ... standard imports
+    from smartsim.ml.tf import serialize_model
+
+    # ... define, instantiate, and compile Keras model
+
+    serialized_model, inputs, outputs = serialize_model(model)
+
+    client = Client(cluster=False)
+    model_key = "tf_mnist_serialized"
+    client.set_model(
+        model_key, serialized_model, "TF", device="GPU", inputs=inputs, outputs=outputs
+    )
+    mnist_image = np.random.rand(1, 28, 28).astype(np.float32)
+    client.put_tensor("mnist_input", mnist_image)
+
+    client.run_model(model_key, "mnist_input", "mnist_output_serialized")
+    pred = client.get_tensor("mnist_output_serialized")
+
+
+PyTorch
+-------
+PyTorch requires models to be `jit-traced <https://pytorch.org/docs/1.11/generated/torch.jit.save.html>`__.
+The method ``torch.jit.save`` can either store the model in memory or on file.
+
+First, we define the model, which in this case is a simple Neural Network:
+
+.. code-block:: python
+
+    import io
+
+    import numpy as np
+    import torch
+    import torch.nn as nn
+    import torch.nn.functional as F
+    from smartredis import Client
+
+    # simple MNIST in PyTorch
+    class Net(nn.Module):
+        def __init__(self):
+            super(Net, self).__init__()
+            self.conv1 = nn.Conv2d(1, 32, 3, 1)
+            self.conv2 = nn.Conv2d(32, 64, 3, 1)
+            self.dropout1 = nn.Dropout(0.25)
+            self.dropout2 = nn.Dropout(0.5)
+            self.fc1 = nn.Linear(9216, 128)
+            self.fc2 = nn.Linear(128, 10)
+
+        def forward(self, x):
+            x = self.conv1(x)
+            x = F.relu(x)
+            x = self.conv2(x)
+            x = F.relu(x)
+            x = F.max_pool2d(x, 2)
+            x = self.dropout1(x)
+            x = torch.flatten(x, 1)
+            x = self.fc1(x)
+            x = F.relu(x)
+            x = self.dropout2(x)
+            x = self.fc2(x)
+            output = F.log_softmax(x, dim=1)
+            return output
+
+We can then create the Neural Network, jit-trace it and upload it
+to the DB. Note that we are storing the serialized model in a ``BytesIO``
+object, which means that we are keeping it in memory and not storing
+it on the file system. For this reason, we need to call SmartRedis's
+``Client.set_model()`` method.
+
+
+.. code-block:: python
```

Comment:
it looks like youre setting the model here from memory and from file, do you mind offering subsubsections here?